### PR TITLE
Cobuild refactor

### PR DESCRIPTION
**Pull request by xxuejie**

### CODE REVIEW

File: c/cobuild.c [hunk 1]
```diff
@@ -879,23 +744,22 @@ int ckb_cobuild_entry(ScriptEntryType callback, bool *cobuild_enabled) {
...
-                                 MAX_CACHE_SIZE, index, CKB_SOURCE_INPUT);
-    // step 6, not WitnessLayoutOtx
-    CHECK_LOOP(err);
-    uint32_t id = 0;
-    err = try_union_unpack_id(&cursor, &id);
-    if (err || id != WitnessLayoutOtx) {
+  for (; index < witness_len; index++) {
+    WitnessLayoutType witness_layout = {0};
+    err = get_witness_layout(witnesses, index, &witness_layout);
+    if (err != 0) {
```
Comment (XuJiandong) [inline c/cobuild.c:757]: Use CHECK_LOOP, in case there is any error inside get_wintess_layout

**Review by mohanson — APPROVED** [review]
Please fix CI.

**Comment by xxuejie** [comment]
I managed to fix the blake2b redefinition issue, however there is a bigger problem: the simulator infrastructure does not provide an implementation of ckb_load_transaction.

It does not seem that `ckb_syscall_omni_lock_sim.h` provides an implementation of `ckb_load_transaction`, nor there is an easy path to add one.

I have to ask: how important is the simulation test? If it is important, shall we migrate to a proper architecture than the **just enough** one here? Or if it is not so critical, can we disable it first and come back to it later?

**Comment by XuJiandong** [comment]
> I managed to fix the blake2b redefinition issue, however there is a bigger problem: the simulator infrastructure does not provide an implementation of ckb_load_transaction.
> 
> It does not seem that `ckb_syscall_omni_lock_sim.h` provides an implementation of `ckb_load_transaction`, nor there is an easy path to add one.
> 
> I have to ask: how important is the simulation test? If it is important, shall we migrate to a proper architecture than the **just enough** one here? Or if it is not so critical, can we disable it first and come back to it later?

It's not important since we have a lot of Rust tests. Simply disable it in CI. We will look into this issue if needed.

File: c/omni_lock.c [hunk 2]
```diff
@@ -137,41 +137,27 @@ bool is_memory_enough(mol_seg_t seg, const uint8_t *cur, uint32_t len) {
...
-  script_seg.size = (mol_num_t)len;
-
-  mol_errno mol_err = MolReader_Script_verify(&script_seg, false);
-  CHECK2(mol_err == MOL_OK, ERROR_ENCODING);
 
-  mol_seg_t args_seg = MolReader_Script_get_args(&script_seg);
-  mol_seg_t seg = MolReader_Bytes_raw_bytes(&args_seg);
-
-  uint8_t *cur = seg.ptr;
+  // TODO: do we need to validate Script structure here?
```
Comment (XuJiandong) [inline c/omni_lock.c:146]: The script structure can be safely used without validation. It's directly from syscall. If this structure is wrong, the ckb consensus is in trouble.

**Review by XuJiandong — APPROVED** [review]

File: c/cobuild.c (hunk 1 above)
Comment (xxuejie) [reply c/cobuild.c:757]: Note we are only doing iteration here within `witness_len`, right now no syscall would throw `CKB_OUT_OF_BOUND` error, meaning `CHECK_LOOP` does not apply here.

Any errors returned by `get_witness_layout` will be essentially due to misformed witness format, personally I think we can safely ignore errors from `get_witness_layout` here.

**Comment by XuJiandong** [comment]
LGTM. We will add following test cases:
1. test cases for CKB without any witness should fail. To test in case witness is absent.
2. test cases with large cell data, witness and script, especially for non-cobuild cases. 
3. Add verify_WitnessLayout/verify_WitnessArgs
4. Add test cases for header deps

**Pull request merged**